File: 240826/최소 차이로 묶기/bundling-with-minimum-difference.py
INF = 10**9 + 1
n, m = map(int, input().split())

arrn = [*map(int,input().split())] #무조건 n이 m보다 크거나 같게 만든다
arrm = [*map(int,input().split())]

if m > n:
    n,m = m,n
    tmp = arrn[::]
    arrn = arrm[::]
    arrm = tmp[::]#깊은 복사
arrn.sort(reverse=1)
arrm.sort(reverse=1)

dp = [[INF]*(n)for _ in range(m)]


def dfs(depth, index):#해당 depth에 index를 골랐을 때 최대 값
    if depth == m:
        return 0
    if index == n:return INF
    if dp[depth][index]!=INF:return dp[depth][index]

    for j in range(index, n):#depth랑 j랑 차이 계산 후 더해주기
        dp[depth][index] = min(dp[depth][index], dfs(depth+1,j+1) + abs(arrm[depth] - arrn[j]))
    return dp[depth][index]
print(dfs(0,0))

File: 240826/최소 차이로 묶기/bundling-with-minimum-difference.cpp
#include <bits/stdc++.h>
using namespace std;
int arrn[1002],arrm[1002];
int dp[1002][1002];
const int INF = 0x3f3f3f3f;
int n,m;

int dfs(int depth, int index){
    if(depth == min(n,m))return 0;
    if(index >= max(n,m))return INF;
    if (dp[depth][index]!=INF)return dp[depth][index];
    
    auto idx = lower_bound(arrm + index, arrm + m, arrn[depth]) - arrm;
    for(int i=index;i<idx;i++){
        dp[depth][index] = min(dp[depth][index], dfs(depth+1,i+1) + abs(arrn[depth] - arrm[i]));
    }
    return dp[depth][index];
}
int main() {
    cin >> n >> m;
    memset(dp, 0x3f, sizeof dp);
    for(int i=0;i<n;i++)cin >> arrn[i];
    for(int j=0;j<m;j++)cin >> arrm[j];

    sort(arrn,arrn+n, greater<int>());
    sort(arrm,arrm+m, greater<int>());
    cout << dfs(0,0);

    return 0;

}